app/admin/tenants/tenants.py

```diff
@@ -0,0 +1,262 @@
+import logging
+from secrets import token_urlsafe
+from typing import List
+from uuid import uuid4
+
+from aries_cloudcontroller import (
+    AcaPyClient,
+    CreateWalletRequest,
+    CreateWalletTokenRequest,
+    RemoveWalletRequest,
+    UpdateWalletRequest,
+)
+from aries_cloudcontroller.model.wallet_record import WalletRecord
+from fastapi import APIRouter, Depends
+
+from app.admin.tenants.models import (
+    CreateTenantRequest,
+    CreateTenantResponse,
+    Tenant,
+    TenantAuth,
+    UpdateTenantRequest,
+    tenant_from_wallet_record,
+)
+from app.admin.tenants.onboarding import (
+    handle_ecosystem_tenant_update,
+    onboard_ecosystem_tenant,
+)
+from app.dependencies import AcaPyAuth, Role, acapy_auth, agent_role
+from app.error import CloudApiException
+from app.facades.trust_registry import (
+    Actor,
+    actor_by_id,
+    register_actor,
+    remove_actor_by_id,
+)
+from app.role import AgentType
+
+logger = logging.getLogger(__name__)
+
+router = APIRouter(prefix="/admin/tenants", tags=["admin: tenants"])
+
+multitenant_admin = agent_role([Role.MEMBER_ADMIN, Role.ECOSYSTEM_ADMIN])
+
+
+def tenant_api_key(role: Role, tenant_token: str):
+    "Get the cloud api key for a tenant with specified role."
+
+    if not role.agent_type.tenant_role:
+        raise CloudApiException("Invalid role", 403)
+
+    return f"{role.agent_type.tenant_role.name}.{tenant_token}"
+
+
+async def get_tenant_for_admin_role(
+    aries_controller: AcaPyClient, tenant_id: str, role: Role
+) -> WalletRecord:
+    """Get the wallet record for a tenant with specified id.
+
+    Will throw an error if the wallet is not created with same role as
+    role specified in method parameters. This is identified by a role prefix
+    to the wallet name. E.g. wallet name "member.xxxxxxxxxxxxxxxx" means
+    the wallet was created in the context of a member.
+
+    Args:
+        aries_controller (AcaPyClient): aries controller to use
+        tenant_id (str): Tenant id to retrieve the wallet for
+        role (Role): Role the wallet should be in
+
+    Raises:
+        CloudApiException: When the role of the wallet does not match the
+            role from the method parameteres
+
+    Returns:
+        WalletRecord: The wallet record.
+    """
+
+    # We retrieve the wallet to make sure it exists
+    wallet = await aries_controller.multitenancy.get_wallet(wallet_id=tenant_id)
+
+    if not wallet_is_for_admin_role(wallet, role):
+        raise CloudApiException(f"Tenant with id {tenant_id} not found", 404)
+
+    return wallet
+
+
+def wallet_is_for_admin_role(wallet: WalletRecord, role: Role):
+    wallet_name: str = wallet.settings.get("wallet.name")
+
+    tenant_agent_type = role.agent_type.tenant_role
+
+    return wallet_name.startswith(f"{tenant_agent_type.name}.")
+
+
+def get_wallet_name_for_role(agent_type: AgentType) -> str:
+    return f"{agent_type.name}.{uuid4().hex}"
```

Comment:
We need to store whether a tenant was created as member or as ecosystem partner. As the cloud api is stateless the easiest way to do this is by prefixing the wallet name with either `member` or `ecosystem`. This is how we can also know whether an admin is allowed to retrieve a wallet by a specific id (ecosystem admin can't retrieve member wallet).

This is only an issue because the ecosystem and member wallets are both created in the same agent, while from the perspective of the cloud api they are separate roles with separate agent endpoints

Not sure if this makes sense 😅